[SPARK-36620][SHUFFLE] Add Push Based Shuffle client side read metrics

thejdeep opened the pull request:
### What changes were proposed in this pull request?
 This adds the following push based shuffle read metrics like :
 - Merger count and Push Based Shuffle enabled/disabled for stage
 - Time spent on results finalization
 - Counting actual number of blocks other than chunks
 - Corrupt shuffle blocks chunks and fallback

 ### Why are the changes needed?
 These changes help to understand the push based shuffle read metrics of the application

 ### Does this PR introduce _any_ user-facing change?
 Changes to API responses by SHS (eg: /stages)

 ### How was this patch tested?
 Modified existing unit tests and also tested API response on event log files in SHS

AmplabJenkins commented:
Can one of the admins verify this patch?

otterc commented on core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:
```diff
@@ -627,9 +709,22 @@ private[spark] class AppStatusStore(
+        pushBasedShuffleMergedLocalChunksFetched = stage.pushBasedShuffleMergedLocalChunksFetched,
+        pushBasedShuffleMergedRemoteBytesRead = stage.pushBasedShuffleMergedRemoteBytesRead,
+        pushBasedShuffleMergedLocalBytesRead = stage.pushBasedShuffleMergedLocalBytesRead,
+        pushBasedShuffleRemoteReqsDuration = stage.pushBasedShuffleRemoteReqsDuration,
```
This looks off. This shouldn't be prefixed by `pushBasedShuffle`

otterc commented on core/src/main/scala/org/apache/spark/status/LiveEntity.scala:
```diff
@@ -502,9 +527,30 @@ private class LiveStage(var info: StageInfo) extends LiveEntity {
+        metrics.shuffleReadMetrics.pushBased.remoteMergedBlocksBytesRead,
+      pushBasedShuffleMergedLocalBytesRead =
+        metrics.shuffleReadMetrics.pushBased.localMergedBlocksBytesRead,
+      pushBasedShuffleRemoteReqsDuration = metrics.shuffleReadMetrics.remoteReqsDuration,
```
Same here. This is not specific to pushBased shuffle

otterc commented on core/src/main/scala/org/apache/spark/status/LiveEntity.scala:
```diff
@@ -843,9 +936,35 @@ private[spark] object LiveEntityHelpers {
+        m2.shuffleReadMetrics.pushBased.localMergedBlocksBytesRead * mult,
+      m1.shuffleReadMetrics.remoteReqsDuration + m2.shuffleReadMetrics.remoteReqsDuration * mult,
+      m1.shuffleReadMetrics.pushBased.remoteMergedReqsDuration +
+        m2.shuffleReadMetrics.pushBased.remoteMergedChunksFetched * mult,
```
This looks off.

otterc reviewed (commented):
This is a big change. Did one pass.

zhouyejoe commented:
@thejdeep  This PR should have missed the changes required in ShuffleBlockFetchIterator which update the ShuffleMetrics while shuffle fetching is on-going.

thejdeep commented:
@zhouyejoe @otterc Updated the PR with the latest changes, please take a look. Thank you!

zhouyejoe commented:
@thejdeep  Thanks for updating the PR. Yet to review it, but several UTs within HistoryServerSuite failed, can you help double check?

thejdeep commented:
@otterc @zhouyejoe Fixed the failing tests, should be good for another review now. Thanks!

mridulm commented on core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala:
```diff
@@ -1219,7 +1219,12 @@ class ShuffleBlockFetcherIteratorSuite extends SparkFunSuite with PrivateMethodT
             new RuntimeException("forced error"))
         }
```
Also update the other tests, particularly related to push based shuffle to assert on the expected metrics ?
Ideally, the original tests should have done it (if the metrics had been added initially) - and so now we have to retroactively fix them ...

mridulm replied:
We can do this as a follow up, given the size of this PR.
Can you please file a jira for this ? Thanks !

mridulm commented on core/src/test/scala/org/apache/spark/ui/StagePageSuite.scala:
```diff
@@ -78,6 +78,16 @@ class StagePageSuite extends SparkFunSuite with LocalSparkContext {
+        shuffleMergedRemoteBytesRead = 1L,
+        shuffleMergedLocalBytesRead = 1L,
+        shuffleRemoteReqsDuration = 1L,
+        shuffleMergedRemoteReqsDuration = 1L,
```
Any particular reason these are set to `1` ?
Particularly when `shuffleMergersCount` == 0 ?

otterc replied:
We didn't put much thought into it. But `shuffleMergerCount` can be 0 because this stage may not push but this stage can still read shuffle data generated by the previous stage.

mridulm replied:
nit: given inputBytes > 0, no likely :-) But that is not relevant to this test !

mridulm commented on core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala:
```diff
@@ -1780,7 +1802,19 @@ private[spark] object JsonProtocolSuite extends Assertions {
-      |      "Total Records Read" : 0
+      |      "Total Records Read" : 0,
+      |      "Remote Requests Duration": 0,
+      |      "Push Based": {
```
Rename this to "Push Based Shuffle" instead ?
It is within the context of `Shuffle Read Metrics`, but makes it more clear.

Thoughts ?

+CC @otterc, @Ngone51

mridulm replied:
+CC @zhouyejoe

mridulm replied:
Let us do this rename as well. Thoughts ?

mridulm commented on core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala:
```diff
@@ -1653,7 +1663,19 @@ private[spark] object JsonProtocolSuite extends Assertions {
+      |      "Remote Requests Duration": 900,
+      |      "Push Based": {
+      |         "Corrupt Merged Block Chunks" : 100,
+      |         "Fallback Count" : 100,
```
Curious why these are nonzero if push based shuffle was disabled ?
Are these simply values to test serde ?

mridulm commented on core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala:
```diff
@@ -2623,48 +2677,118 @@ private[spark] object JsonProtocolSuite extends Assertions {
       |        {
-      |          "ID": 24,
+      |          "ID": 34,
       |          "Name": "${output.RECORDS_WRITTEN}",
```
Curious why some of these values changed.
I am yet to look at tests more carefully though ...

thejdeep replied:
The values just changed since metrics were added that took the place of indices preceding this.

mridulm reviewed (commented):
Had a few comments, but overall looks good.
Thanks for working on this @thejdeep !

otterc commented on core/src/main/scala/org/apache/spark/storage/PushBasedFetchHelper.scala:
```diff
@@ -290,6 +301,7 @@ private class PushBasedFetchHelper(
       address: BlockManagerId): Unit = {
     assert(blockId.isInstanceOf[ShuffleMergedBlockId] || blockId.isInstanceOf[ShuffleBlockChunkId])
     logWarning(s"Falling back to fetch the original blocks for push-merged block $blockId")
+    shuffleMetrics.incFallbackCount(1)
```
Fallback count measures the number of times fallback is initiated for a block which is either a `shuffleMergedBlock` or a `shuffleBlockChunk`. I don't understand why we need to move into the case that you suggested.

mridulm replied:
I misread the change, makes sense- thanks !

thejdeep commented on core/src/main/scala/org/apache/spark/status/storeTypes.scala:
```diff
@@ -233,6 +243,38 @@ private[spark] class TaskDataWrapper(
+    val shuffleRemoteReqsDuration: Long,
+    @KVIndexParam(
+      value = TaskIndexNames.PUSH_BASED_SHUFFLE_MERGED_REMOTE_REQS_DURATION,
+      parent = TaskIndexNames.STAGE)
```
Indexes for these fields is a requirement when we are computing quantiles and reading `TaskDataWrapper`s.

mridulm replied:
Makes sense, thanks for clarifying !

thejdeep committed:
SPARK-36620 : Push based shuffle read metrics

 # What changes were proposed in this pull request?
This adds the following push based shuffle metrics like :

- Merger count and Push Based Shuffle enabled/disabled for stage
- Time spent on results finalization
- Counting actual number of blocks other than chunks
- Corrupt shuffle blocks chunks and fallback
 # Why are the changes needed?
These changes help to understand the push based shuffle read metrics of the application

 # Does this PR introduce any user-facing change?
Changes to API responses by SHS (eg: /stages)

 # How was this patch tested?
Modified existing unit tests and also tested API response on event log files in SHS

thejdeep commented on core/src/main/scala/org/apache/spark/executor/Executor.scala:
```diff
@@ -654,6 +654,27 @@ private[spark] class Executor(
+        executorSource.METRIC_PUSH_BASED_SHUFFLE_REMOTE_REQS_DURATION
+          .inc(task.metrics.shuffleReadMetrics.remoteReqsDuration)
+        executorSource.METRIC_PUSH_BASED_SHUFFLE_MERGED_REMOTE_REQS_DURATION
+          .inc(task.metrics.shuffleReadMetrics.remoteMergedReqsDuration)
```
Added method to populate these.

mridulm commented:
+CC @zhouyejoe

zhouyejoe commented on core/src/main/scala/org/apache/spark/executor/ShuffleReadMetrics.scala:
```diff
@@ -146,6 +268,16 @@ private[spark] class TempShuffleReadMetrics extends ShuffleReadMetricsReporter {
+  override def incLocalMergedChunksFetched(v: Long): Unit = _localMergedChunksFetched += v
+  override def incRemoteMergedBlocksBytesRead(v: Long): Unit = _remoteMergedBlocksBytesRead += v
+  override def incLocalMergedBlocksBytesRead(v: Long): Unit = _localMergedBlocksBytesRead += v
+  override def incRemoteReqsDuration(v: Long): Unit = _remoteReqsDuration += v
+  override def incRemoteMergedReqsDuration(v: Long): Unit = _remoteMergedReqsDuration += v
```
I am able to find the calling for these two methods in SHS which can build the metrics up from JSON, but the caller in the ShuffleBlockFetcherIterator is actually missing in this patch. For the caller part in Executor, which measures the remote request duration, needs more refactoring in the Iterator. Probably we can carve these two metrics out in this PR. If needed, we can create another PR which tracks the remote request duration for unmerged and merged shuffle. @mridulm @otterc Thoughts?

mridulm replied:
I called out the same [above](https://github.com/apache/spark/pull/36165#discussion_r978229580)
I am fine with pulling those two metrics out and introducing them in a different PR

zhouyejoe commented on core/src/test/scala/org/apache/spark/ui/StagePageSuite.scala:
```diff
@@ -78,6 +78,16 @@ class StagePageSuite extends SparkFunSuite with LocalSparkContext {
         shuffleLocalBytesRead = 1L,
         shuffleReadBytes = 1L,
         shuffleReadRecords = 1L,
+        shuffleCorruptMergedBlockChunks = 2L,
+        shuffleFallbackCount = 2L,
```
Any particular reason these are set to 2?

zhouyejoe commented on core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala:
```diff
@@ -227,6 +227,16 @@ class TaskMetrics private[spark] () extends Serializable {
+    shuffleRead.REMOTE_MERGED_REQS_DURATION -> shuffleReadMetrics._remoteMergedReqsDuration,
     shuffleWrite.BYTES_WRITTEN -> shuffleWriteMetrics._bytesWritten,
     shuffleWrite.RECORDS_WRITTEN -> shuffleWriteMetrics._recordsWritten,
     shuffleWrite.WRITE_TIME -> shuffleWriteMetrics._writeTime,
```
I think we are missing the three push related metrics for the ShuffleWriter, including "BlocksNotPushed", "BlocksTooLate" and  "BlocksCollided" here, and also the other relevant part, including the metrics logging in the Executor, the rendering in SHS and also in the SQLShuffleMetricsReporter.

zhouyejoe replied:
Chatted offline. Ignore my comments on this one since it is planned to do it in a separate PR.

zhouyejoe reviewed (commented):
Took a thorough review, excluding the UTs.
Major comments:
1. writer side push based shuffle metrics are not included in this PR 
2. Suggest to remove the remote request duration metrics in the reader side.

mridulm commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -282,6 +280,17 @@ final class ShuffleBlockFetcherIterator(
+          shuffleMetrics.incRemoteMergedReqsDuration(durationMs)
+        } else {
+          shuffleMetrics.incRemoteReqsDuration(durationMs)
+        }
```
`shuffleLocalMetricsUpdate` and `shuffleRemoteMetricsUpdate` is treating merged fetches as a subset of all fetches - while here we are treating them seperately.
Any particular reason for this difference ?

zhouyejoe replied:
+1 on this point, the merged remote request duration should be counted into the total remote request duration, following the same logic for the local/remote bytes read.
@otterc WDTY?

otterc replied:
We added these metrics internally so we can easily figure out what is taking longer- the fetch of the merge chunks or the few unmerged blocks. If we include the duration of merged chunks in the total then we have to derive the value for un-merged blocks which is fine. It makes sense to be consistent to avoid confusion.

zhouyejoe commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -726,6 +736,63 @@ final class ShuffleBlockFetcherIterator(
+  }
+
+  private def shuffleMetricsUpdate(
+    blockId: BlockId,
+    buf: ManagedBuffer,
+    local: Boolean): Unit = {
```
Nit: Indent is off

zhouyejoe commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -722,6 +718,63 @@ final class ShuffleBlockFetcherIterator(
+        shuffleMetrics.incLocalMergedChunksFetched(1)
+        shuffleMetrics.incLocalMergedBlocksFetched(getShuffleChunkCardinality(chunkId))
+        shuffleMetrics.incLocalMergedBlocksBytesRead(buf.size)
+        shuffleMetrics.incLocalBlocksFetched(getShuffleChunkCardinality(chunkId))
```
getShuffleChunkCardinality will only get called once for each blockId in either shuffleLocalMetricsUpdate or shuffleRemoteMetricsUpdate, I think it won't get reused.

mridulm replied:
Take a look at the snippet above - it is used twice :-)
Same in `shuffleRemoteMetricsUpdate`

zhouyejoe commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -407,6 +407,8 @@ private[spark] object JsonProtocol {
     g.writeFieldName("Accumulables")
     accumulablesToJson(stageInfo.accumulables.values, g)
     g.writeNumberField("Resource Profile Id", stageInfo.resourceProfileId)
+    g.writeBooleanField("Push based shuffle enabled", stageInfo.isPushBasedShuffleEnabled)
+    g.writeNumberField("Shuffle push mergers count", stageInfo.shuffleMergerCount)
```
Nit: Capitalization for the first char of each word, following the same style as others shown above

zhouyejoe commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -1105,11 +1132,17 @@ private[spark] object JsonProtocol {
     }
+    val isPushBasedShuffleEnabled =
+      jsonOption(json.get("Push based shuffle enabled")).map(_.extractBoolean).getOrElse(false)
+    val shufflePushMergersCount =
```
Nit: Same here. Capitalization

zhouyejoe commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -1226,6 +1259,29 @@ private[spark] object JsonProtocol {
+      jsonOption(readJson.get("Push Based")).foreach { pushBasedShuffleReadJson =>
+        readMetrics.incCorruptMergedBlockChunks(jsonOption(
+          pushBasedShuffleReadJson.get("Corrupt Merged Block Chunks"))
+          .map(_.extractLong).getOrElse(0L))
```
Nit: Looks like indent is off

zhouyejoe commented on core/src/test/resources/HistoryServerExpectations/one_stage_json_with_details_expectation.json:
```diff
@@ -32,7 +32,19 @@
+  "shuffleMergedRemoteBytesRead" : 0,
+  "shuffleMergedLocalBytesRead" : 0,
+  "shuffleRemoteReqsDuration" : 0,
+  "shuffleMergedRemoteReqsDuration" : 0,
   "shuffleReadRecords" : 0,
+  "isPushBasedShuffleEnabled" : false,
```
Nit: Line 45 should be moved to line 35.

zhouyejoe commented on project/MimaExcludes.scala:
```diff
@@ -87,6 +87,11 @@ object MimaExcludes {
+    ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.status.api.v1.ShuffleReadMetrics.this"),
+    ProblemFilters.exclude[MissingMethodProblem]("org.apache.spark.status.api.v1.StageData.this"),
+
     // [SPARK-40324][SQL] Provide query context in AnalysisException
```
Just curious: What specific changes lead to this exclusion? Can we avoid this exclusion?

thejdeep replied:
There are some binary incompatibilities that had to be added to the excludes list.

zhouyejoe reviewed (commented):
There are a lot of 0 values for push based shuffle in multiple JSON files under resources/HistoryServerExpectations. The main reason is for fixing the schema issue as we introduced new records in the stages rest call response. But this only tests that schema changes works as expected, but not testing whether SHS can parse the log files and respond with the expected value? I think we can try to change some of the event files under spark-events folder, which includes some actual metrics for push based shuffle, which will also require changes in the expected JSON response. Thus we not only test the schema change, but also test SHS works out on parsing and rendering the JSON response after parsing the log events.

venkata91 committed:
Author:    Venkata krishnan Sowrirajan <vsowrirajan@linkedin.com>

otterc committed:
Author:    Chandni Singh <chsingh@linkedin.com>

zhouyejoe committed:
Author:    Ye Zhou <yezhou@linkedin.com>

xkrogen committed:
Author:    Erik Krogen <ekrogen@linkedin.com>

thejdeep commented:
Thanks for reviewing @zhouyejoe @mridulm, addressed feedback.

mridulm commented:
Please don't force push the code, unless it is to resolve conflicts/etc which can be handled otherwise - it makes reviewing what changed much harder

mridulm commented on core/src/main/scala/org/apache/spark/status/storeTypes.scala:
```diff
@@ -138,6 +138,16 @@ private[spark] object TaskIndexNames {
+  final val PUSH_BASED_SHUFFLE_MERGED_LOCAL_CHUNKS = "pbsmlc"
+  final val PUSH_BASED_SHUFFLE_MERGED_REMOTE_READS = "pbsmrr"
+  final val PUSH_BASED_SHUFFLE_MERGED_LOCAL_READS = "pbsmlr"
+  final val PUSH_BASED_SHUFFLE_REMOTE_REQS_DURATION = "pbsrrd"
```
This is not specific to push based shuffle - `srrd` instead ?

mridulm commented on core/src/main/scala/org/apache/spark/executor/Executor.scala:
```diff
@@ -791,6 +770,53 @@ private[spark] class Executor(
+      executorSource.METRIC_PUSH_BASED_SHUFFLE_MERGED_LOCAL_BLOCKS_FETCHED
+        .inc(metrics.shuffleReadMetrics.localMergedBlocksBytesRead)
+      executorSource.METRIC_PUSH_BASED_SHUFFLE_REMOTE_REQS_DURATION
+        .inc(metrics.shuffleReadMetrics.remoteReqsDuration)
```
This is not specific to push based shuffle, right ? Why is the metric prefixed with that namespace ? (Here and in other places where this metric is referenced)

mridulm commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -726,6 +736,65 @@ final class ShuffleBlockFetcherIterator(
 
+  // Number of map blocks in a ShuffleBlockChunk
+  private def getShuffleChunkCardinality(blockId: ShuffleBlockChunkId): Int = {
+    val chunkTracker = pushBasedFetchHelper.getRoaringBitMap(blockId)
+    chunkTracker match {
+      case Some(bitmap) => bitmap.getCardinality
+      case None => 0
+    }
```
```suggestion
    pushBasedFetchHelper.getRoaringBitMap(blockId).map(_.getCardinality).getOrElse(0)
```

mridulm commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -866,6 +930,7 @@ final class ShuffleBlockFetcherIterator(
                 }
 
                 if (blockId.isShuffleChunk) {
+                  shuffleMetrics.incCorruptMergedBlockChunks(1)
```
We need to add this for the invocation of `initiateFallbackFetchForPushMergedBlock` [above](https://github.com/apache/spark/blob/72cce5c39da8a52efa0a7cad4d65dac4eb389ff6/core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala#L806) as well - the zero byte read is unexpected, and is essentially a corrupt read

thejdeep replied:
Thanks for calling this out, added this change.

mridulm reviewed (commented):
Added a few comments.

thejdeep committed:
Update protobuf serializer and add store types for push shuffle client read metrics

thejdeep commented on core/src/main/scala/org/apache/spark/InternalAccumulator.scala:
```diff
@@ -54,6 +55,16 @@ private[spark] object InternalAccumulator {
     val RECORDS_READ = SHUFFLE_READ_METRICS_PREFIX + "recordsRead"
+    val REMOTE_REQS_DURATION = SHUFFLE_READ_METRICS_PREFIX + "remoteReqsDuration"
+    val CORRUPT_MERGED_BLOCK_CHUNKS = SHUFFLE_PUSH_READ_METRICS_PREFIX + "corruptMergedBlockChunks"
+    val FALLBACK_COUNT = SHUFFLE_PUSH_READ_METRICS_PREFIX + "fallbackCount"
```
renamed in all sources to `mergedFetchFallbackCount `

zhouyejoe commented on sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLShuffleMetricsReporter.scala:
```diff
@@ -44,6 +44,26 @@ class SQLShuffleReadMetricsReporter(
+  private[this] val _corruptMergedBlockChunks =
+    metrics(SQLShuffleReadMetricsReporter.CORRUPT_MERGED_BLOCK_CHUNKS)
+  private[this] val _fallbackCount =
+    metrics(SQLShuffleReadMetricsReporter.FALLBACK_COUNT)
```
Build failed on this line. Need rename it to MERGED_FETCH_FALLBACK_COUNT.

mridulm commented on core/src/main/scala/org/apache/spark/executor/Executor.scala:
```diff
@@ -791,6 +770,53 @@ private[spark] class Executor(
+      executorSource.METRIC_PUSH_BASED_SHUFFLE_CORRUPT_MERGED_BLOCK_CHUNKS
+        .inc(metrics.shuffleReadMetrics.corruptMergedBlockChunks)
+      executorSource.METRIC_PUSH_BASED_SHUFFLE_MERGED_FETCH_FALLBACK_COUNT
+        .inc(metrics.shuffleReadMetrics.fallbackCount)
```
Rename `fallbackCount` (and corresponding `_fallbackCount`) to match the metric name ?
`mergedFetchFallbackCount` like in other places ?

mridulm commented on core/src/main/scala/org/apache/spark/status/storeTypes.scala:
```diff
@@ -233,6 +243,38 @@ private[spark] class TaskDataWrapper(
+    val shuffleMergedLocalChunksFetched: Long,
+    @KVIndexParam(
+      value = TaskIndexNames.PUSH_BASED_SHUFFLE_MERGED_REMOTE_READS, parent = TaskIndexNames.STAGE)
+    val shuffleMergedRemoteBlocksBytesRead: Long,
```
`Blocks` is not required for `BytesRead` - change here and in all other similar cases.
For ex: `shuffleMergedRemoteBlocksBytesRead` -> `shuffleMergedRemoteBytesRead`, `shuffleMergedLocalBlocksBytesRead` -> `shuffleMergedLocalBytesRead`

Same for `ShufflePushReadMetrics`, etc.

mridulm commented on core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:
```diff
@@ -319,7 +319,18 @@ private[spark] class AppStatusStore(
+          toValues(_.shuffleRemoteReqsDuration),
+          new v1.ShufflePushReadMetricDistributions(
+            toValues(_.shuffleCorruptMergedBlockChunks),
+            toValues(_.shuffleFallbackCount),
```
`shuffleFallbackCount` and `ShufflePushReadMetricDistributions.fallbackCount` needs to be renamed as well to be consistent.

Please do make sure the other renames we did are consistent throughout as well - I will go over it once more later, but would be great if you can take a pass as well @thejdeep . Thx

mridulm commented on sql/core/src/main/scala/org/apache/spark/sql/execution/metric/SQLShuffleMetricsReporter.scala:
```diff
@@ -44,6 +44,26 @@ class SQLShuffleReadMetricsReporter(
     metrics(SQLShuffleReadMetricsReporter.FETCH_WAIT_TIME)
   private[this] val _recordsRead =
     metrics(SQLShuffleReadMetricsReporter.RECORDS_READ)
```
Keep metrics here consistent with the other renames as well.

mridulm commented on core/src/test/resources/HistoryServerExpectations/excludeOnFailure_node_for_stage_expectation.json:
```diff
@@ -81,7 +93,19 @@
           "localBytesRead" : 0,
-          "recordsRead" : 0
+          "recordsRead" : 0,
+          "pushBased": {
```
This should be "Push Based shuffle" right ?

thejdeep replied:
This is the API response that serialized based on the [case classes in api.scala here](https://github.com/apache/spark/blob/7259a8c0147d9da70848ef2e7499a1712a4c4741/core/src/main/scala/org/apache/spark/status/api/v1/api.scala#L384). We only write "Push Based Shuffle" when logging to the event file

thejdeep replied:
Do you recommend changing this to "pushBasedShuffle" ? to note:  it is inside the context of shuffle read metrics)

mridulm replied:
Yes, that would be better.

thejdeep replied:
Changed it to "pushBasedShuffle", thanks for the suggestion.

mridulm reviewed (commented):
Just one comment, rest looks good to me.

mridulm reviewed (approved):
Looks good to me.
Thanks for all the changes @thejdeep !

+CC @Ngone51, @otterc, @zhouyejoe 

Will give others some time to take a look at the PR

otterc reviewed (commented):
Partial review.

otterc commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -726,6 +736,61 @@ final class ShuffleBlockFetcherIterator(
+  }
+
+  private def shuffleLocalMetricsUpdate(blockId: BlockId, buf: ManagedBuffer): Unit = {
+    // Check if the block is within the host-local blocks set, or if it is a merged local
```
Where is this check? Do we need to rephrase this comment?

otterc reviewed (commented):
Need to verify these.

otterc commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -726,6 +736,61 @@ final class ShuffleBlockFetcherIterator(
+
+  // Check if the merged block is local to the host or not
+  private def isLocalMergedBlockAddress(address: BlockManagerId): Boolean = {
+    address.executorId.isEmpty && address.host == blockManager.blockManagerId.host
```
This method is there in PushBasedFetchHelper. `PushBasedFetchHelper.isLocalPushMergedBlockAddress`. We should try to just use that.  We can maybe make that static. The executor id is not empty. It's `SHUFFLE_MERGER_IDENTIFIER`

mridulm reviewed (commented):
Please address @otterc's comments @thejdeep.
I will take a pass over it once they are done. Thanks !

mridulm commented on core/src/main/scala/org/apache/spark/status/api/v1/api.scala:
```diff
@@ -302,7 +312,9 @@ class StageData private[spark](
     val taskMetricsDistributions: Option[TaskMetricDistributions],
-    val executorMetricsDistributions: Option[ExecutorMetricsDistributions])
+    val executorMetricsDistributions: Option[ExecutorMetricsDistributions],
+    val isPushBasedShuffleEnabled: Boolean,
```
We actually have a lot of `pushBased*` references in the codebase.
Given the block id's are named shufflePush, that might be a good prefix to settle on.

thejdeep replied:
Removed `Based` from user facing metrics.

otterc commented on core/src/test/resources/HistoryServerExpectations/one_stage_json_with_partitionId_expectation.json:
```diff
@@ -178,12 +212,24 @@
           "remoteBytesReadToDisk" : 0,
-          "localBytesRead" : 372,
-          "recordsRead" : 9
+          "localBytesRead" : 0,
+          "recordsRead" : 0,
```
same here

otterc reviewed (commented):
finished review

mridulm reviewed (changes requested):
Please address @otterc's comments @thejdeep.
I will take a pass over it once they are done. Thanks !

thejdeep commented on core/src/test/resources/HistoryServerExpectations/failed_stage_list_json_expectation.json:
```diff
@@ -27,13 +27,23 @@
-  "shuffleLocalBlocksFetched" : 64,
-  "shuffleFetchWaitTime" : 1,
+  "shuffleLocalBlocksFetched" : 0,
+  "shuffleFetchWaitTime" : 0,
```
There was a logical bug in merging the read metrics, thanks for catching this! Updated the expectation JSON.

otterc commented on core/src/main/scala/org/apache/spark/status/LiveEntity.scala:
```diff
@@ -843,6 +917,25 @@ private[spark] object LiveEntityHelpers {
+        m2.shuffleReadMetrics.shufflePushReadMetrics.localMergedBytesRead * mult,
+      m1.shuffleReadMetrics.remoteReqsDuration + m2.shuffleReadMetrics.remoteReqsDuration * mult,
+      m1.shuffleReadMetrics.shufflePushReadMetrics.remoteMergedReqsDuration +
+        m2.shuffleReadMetrics.shufflePushReadMetrics.remoteMergedChunksFetched * mult,
```
This should be `remoteMergedReqsDuration` instead of `remoteMergedChunksFetched`

otterc commented on core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:
```diff
@@ -726,6 +736,56 @@ final class ShuffleBlockFetcherIterator(
     }
   }
 
+  // Number of map blocks in a ShuffleBlockChunk
+  private def getShuffleChunkCardinality(blockId: ShuffleBlockChunkId): Int = {
+    pushBasedFetchHelper.getRoaringBitMap(blockId).map(_.getCardinality).getOrElse(0)
+  }
```
Should we move this to `PushBasedFetchHelper` since most of the utils belonging to pushShuffle are there?

otterc commented on core/src/test/resources/HistoryServerExpectations/stage_with_peak_metrics_expectation.json:
```diff
@@ -33,6 +33,16 @@
+  "shuffleMergedLocalChunksFetched" : 160,
+  "shuffleMergedRemoteBytesRead" : 1600,
+  "shuffleMergedLocalBytesRead" : 1600,
+  "shuffleRemoteReqsDuration" : 0,
```
if mergedRemoteReqDuration > 0, then shuffleRemoteReqsDuration should be greater than mergedRemoteReqDuration. Can we update this?
The existing values for shuffleLocalBytesRead, and shuffleReadBytes  also don't make sense now because if we are saying that there were merged local blocks, then these value should be > 0. 
We can either change or ignore. I am good with either.

otterc commented on core/src/test/resources/HistoryServerExpectations/stage_with_summaries_expectation.json:
```diff
@@ -33,6 +33,16 @@
+  "shuffleMergedLocalChunksFetched" : 160,
+  "shuffleMergedRemoteBytesRead" : 1600,
+  "shuffleMergedLocalBytesRead" : 1600,
+  "shuffleRemoteReqsDuration" : 0,
```
Same here

otterc commented on core/src/test/resources/HistoryServerExpectations/stage_with_summaries_expectation.json:
```diff
@@ -1094,5 +1308,7 @@
+  },
+  "isShufflePushEnabled" : false,
+  "shuffleMergersCount" : 0
+}
```
Missing new line at the end of file

otterc commented on core/src/test/resources/spark-events/app-20200706201101-0003:
```diff
@@ -100,25 +100,25 @@
 {"Event":"SparkListenerJobEnd","Job ID":2,"Completion Time":1594091483044,"Job Result":{"Result":"JobSucceeded"}}
 {"Event":"org.apache.spark.sql.execution.ui.SparkListenerSQLExecutionEnd","executionId":0,"time":1594091483045}
-{"Event":"SparkListenerApplicationEnd","Timestamp":1594091824231}
+{"Event":"SparkListenerApplicationEnd","Timestamp":1594091824231}
```
Missing new line at the end of file

otterc commented on core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala:
```diff
@@ -1653,7 +1663,19 @@ private[spark] object JsonProtocolSuite extends Assertions {
+      |         "Merged Local Chunks Fetched" : 0,
+      |         "Merged Remote Bytes Read" : 0,
+      |         "Merged Local Bytes Read" : 0,
+      |         "Merged Remote Requests Duration": 900
```
If merged remote blocks are 0 then this can't be > 0

thejdeep replied:
This is [currently being set](https://github.com/apache/spark/blob/e46031651617393bb182e44ae04920d6c89592f6/core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala#L1315) to `executorDeserializeTime` + `jvmGCTime` while making the test task metrics, do you think that assumption is incorrect ?

otterc replied:
merge remote request duration is not at all related to `executorDeserializeTime` and `jvmGCTime`. I guess we didn't pay attention to these tests internally.

otterc commented on project/MimaExcludes.scala:
```diff
@@ -87,6 +87,11 @@ object MimaExcludes {
     // [SPARK-36511][MINOR][SQL] Remove ColumnIOUtil
     ProblemFilters.exclude[MissingClassProblem]("org.apache.parquet.io.ColumnIOUtil"),
 
+    // [SPARK-36620] [SHUFFLE] Expose push based shuffle metrics
```
Reword to `Expose shuffle push metrics`

otterc reviewed (commented):
Have some minor comments. Looks good to me

zhouyejoe commented:
There are also UT failures related to this PR @thejdeep

thejdeep commented:
Addressed comments, please take another look @otterc @mridulm @zhouyejoe 
Thanks!

otterc reviewed (approved):
Looks good to me. Thanks for addressing all the comments.

otterc commented:
@thejdeep The test failures are related to this change. Please fix them.

```
/home/runner/work/spark/spark/core/src/main/scala/org/apache/spark/status/protobuf/TaskDataWrapperSerializer.scala:92:5: not enough arguments for constructor TaskDataWrapper: (taskId: Long, index: Int, attempt: Int, partitionId: Int, launchTime: Long, resultFetchStart: Long, duration: Long, executorId: String, host: String, status: String, taskLocality: String, speculative: Boolean, accumulatorUpdates: Seq[org.apache.spark.status.api.v1.AccumulableInfo], errorMessage: Option[String], hasMetrics: Boolean, executorDeserializeTime: Long, executorDeserializeCpuTime: Long, executorRunTime: Long, executorCpuTime: Long, resultSize: Long, jvmGcTime: Long, resultSerializationTime: Long, memoryBytesSpilled: Long, diskBytesSpilled: Long, peakExecutionMemory: Long, inputBytesRead: Long, inputRecordsRead: Long, outputBytesWritten: Long, outputRecordsWritten: Long, shuffleRemoteBlocksFetched: Long, shuffleLocalBlocksFetched: Long, shuffleFetchWaitTime: Long, shuffleRemoteBytesRead: Long, shuffleRemoteBytesReadToDisk: Long, shuffleLocalBytesRead: Long, shuffleRecordsRead: Long, shuffleCorruptMergedBlockChunks: Long, shuffleMergedFetchFallbackCount: Long, shuffleMergedRemoteBlocksFetched: Long, shuffleMergedLocalBlocksFetched: Long, shuffleMergedRemoteChunksFetched: Long, shuffleMergedLocalChunksFetched: Long, shuffleMergedRemoteBytesRead: Long, shuffleMergedLocalBytesRead: Long, shuffleRemoteReqsDuration: Long, shuffleMergedRemoteReqDuration: Long, shuffleBytesWritten: Long, shuffleWriteTime: Long, shuffleRecordsWritten: Long, stageId: Int, stageAttemptId: Int)org.apache.spark.status.TaskDataWrapper.
[error] Unspecified value parameters shuffleCorruptMergedBlockChunks, shuffleMergedFetchFallbackCount, shuffleMergedRemoteBlocksFetched...
[error]     new TaskDataWrapper(
[error]     ^
```

thejdeep commented:
@otterc @mridulm The test failure is from a rebased commit. I had to rebase locally and force push, this rewrote commit history.

mridulm commented on core/src/main/protobuf/org/apache/spark/status/protobuf/store_types.proto:
```diff
@@ -100,11 +100,21 @@ message TaskDataWrapper {
+  int64 shuffle_write_time = 48;
+  int64 shuffle_records_written = 49;
+  int64 stage_id = 50;
+  int32 stage_attempt_id = 51;
```
We should not change attribute `id`'s (here and elsewhere)
In this specific case, given this is not yet released, it is not as bad - but it might still interfere with some ongoing dev/test work.

thejdeep replied:
@mridulm Do you suggest we keep the attribute `id`s same as before ? This would mean having to change the function definitions across most of the diff in this PR

mridulm replied:
Protobuf id's should never be changed.

thejdeep replied:
@mridulm Thanks for reviewing, changed it to keep the existing protobuf ids

mridulm reviewed (commented):
Looks good to me, barring a comment.
Since @otterc and @zhouyejoe reviewed it, want to make sure they are fine with the change as well.

Also, +CC @Ngone51 if you have the bandwidth :-)

mridulm reviewed (approved):
Looks good to me.
+CC @otterc, @zhouyejoe PTAL

thejdeep commented:
@mridulm `StageDataWrapper` was also added to protobuf serialization in #39192 (merged a few days ago) causing this build to fail. I have updated this PR to make more changes to the protobuf store types and also added serialization for push shuffle metrics. The build should be passing now. Thanks.

otterc reviewed (approved):
Looks good to me. Thanks for addressing all the comments @thejdeep

mridulm commented:
Can you take a look at the test failures @thejdeep ? They look relevant.

mridulm commented on core/src/main/protobuf/org/apache/spark/status/protobuf/store_types.proto:
```diff
@@ -652,6 +699,20 @@ message ShuffleReadMetricDistributions {
+  repeated double remote_merged_bytes_read = 7;
+  repeated double local_merged_bytes_read = 8;
+  repeated double remote_merged_reqs_duration = 9;
 }
```
+CC @gengliangwang 
Can you please take a look specifically at the protobuf definitions and make sure they are aligned with your mental model of how this should be modeled ?

Thanks !

gengliangwang replied:
Yes, the protobuf definitions LGTM

mridulm replied:
Thanks @gengliangwang !

gengliangwang commented on core/src/main/scala/org/apache/spark/status/storeTypes.scala:
```diff
@@ -233,6 +243,38 @@ private[spark] class TaskDataWrapper(
     val shuffleLocalBytesRead: Long,
     @KVIndexParam(value = TaskIndexNames.SHUFFLE_READ_RECORDS, parent = TaskIndexNames.STAGE)
     val shuffleRecordsRead: Long,
+    @KVIndexParam(
```
QQ: Do we have the requirement to sort task data by the new metrics?

thejdeep replied:
Although we do not sort these in the UI, adding indices for these are a requirement for computing quantiles and while fetching `TaskDataWrapper`s.

thejdeep commented:
Fixed failing tests and updated commit messages to reflect the overall changes. PTAL. Thanks

mridulm commented:
I am not sure what has changed since last review due to the force pushes - please do not force push, particularly for large PR's.

mridulm commented on core/src/main/scala/org/apache/spark/status/protobuf/StageDataWrapperSerializer.scala:
```diff
@@ -496,7 +555,52 @@ class StageDataWrapperSerializer extends ProtobufSerDe {
+  }
+
+  private def deserializeShufflePushReadMetricsDistributions(
+      binary: StoreTypes.ShuffleReadMetricDistributions): ShuffleReadMetricDistributions = {
```
Remove this method ? It appears unused (and incorrectly named if it is required)

thejdeep replied:
Removed the duplicate method, thanks.

mridulm reviewed (commented):
Rest looks good, added a comment.

thejdeep committed:
Remove duplicate method

mridulm commented:
Looks good to me, will wait for successful build before merging.

zhouyejoe commented:
LGTM. Thanks for working on this PR. @thejdeep

mridulm commented:
Merged to master.
Thanks for working on this @thejdeep !
Thanks for the reviews @otterc, @zhouyejoe, @gengliangwang :-)

JoshRosen commented:
I think that we should skip logging of these metrics in JsonProtocol when push-based shuffle is disabled, so I've filed https://issues.apache.org/jira/browse/SPARK-42203
